go/consensus/tendermint/full/runtimes.go

```diff
@@ -0,0 +1,333 @@
+func (t *fullService) indexRuntime(tr *trackedRuntime, indexHeight int64) error {
+	var err error
+	runtimeID := tr.runtimeID
+	logger := tr.logger.With("index_height", indexHeight)
+
+	crash.Here(crashPointBlockBeforeIndex)
+
+	// Load last indexed height.
+	var lastHeight int64
+	if lastHeight, err = tr.blockHistory.LastConsensusHeight(); err != nil {
+		return fmt.Errorf("failed to get last indexed height: %w", err)
+	}
+	lastNonSeenHeight := lastHeight + 1
+
+	switch {
+	case lastNonSeenHeight > indexHeight:
+		// Already indexed, skip.
+		logger.Debug("height already indexed, skipping", "last_height", lastHeight)
+		return nil
+	case lastNonSeenHeight == indexHeight:
+		// Nothing to reindex, just index the incoming height bellow.
+	default:
+		// We need to reindex. Adjust last non seen height to last available height.
+		var lastRetainedHeight int64
+		lastRetainedHeight, err = t.GetLastRetainedVersion(t.ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get last retained height: %w", err)
+		}
+		if lastNonSeenHeight < lastRetainedHeight {
+			logger.Debug("last non seen height pruned, skipping until last retained",
+				"last_retained_height", lastRetainedHeight,
+				"last_height", lastNonSeenHeight,
+			)
+			lastNonSeenHeight = lastRetainedHeight
+		}
+
+		// Take initial genesis height into account.
+		var genesisDoc *genesisAPI.Document
+		genesisDoc, err = t.GetGenesisDocument(t.ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get genesis document: %w", err)
+		}
+		if lastNonSeenHeight < genesisDoc.Height {
+			logger.Debug("genesis height greater than last non seen height, skip until genesis height",
+				"genesis_height", genesisDoc.HaltEpoch,
+				"last_height", lastNonSeenHeight,
+			)
+			lastNonSeenHeight = genesisDoc.Height
+		}
+
+	}
+
+	// Index all blocks between last available height and current height.
+	logger.Debug("indexing blocks",
+		"last_non_seen_height", lastNonSeenHeight,
+		logging.LogEvent, api.LogEventHistoryReindexing,
+	)
+
+	for height := lastNonSeenHeight; height <= indexHeight; height++ {
+		// Index runtime events.
+		var stakingEvs []*staking.Event
+		stakingEvs, err = t.getRuntimeStakingEvents(runtimeID, height)
+		if err != nil {
+			logger.Error("querying runtime staking events",
+				"height", height,
+				"err", err,
+			)
+			return fmt.Errorf("failed to get staking events: %w", err)
+		}
+		if err = tr.blockHistory.CommitPendingConsensusEvents(height, stakingEvs); err != nil {
+			logger.Error("committing pending consensus events", "err", err, "num_evens", len(stakingEvs))
+			return fmt.Errorf("failed to set consensus events: %w", err)
+		}
+
+		crash.Here(crashPointBlockAfterConsensusEventsIndex)
+
+		var annBlk *api.AnnotatedBlock
+		var roundResults *api.RoundResults
+		annBlk, roundResults, err = t.getRuntimeRoothashEvents(runtimeID, height)
+		if err != nil {
+			return fmt.Errorf("failed to query roothash events: %w", err)
+		}
+		if annBlk == nil || roundResults == nil {
```

Comment:
If either is nil? Shouldn't they be consistent? What if one is nil and the other one is not, will things get missed?